documentation/CAMARA-Security-Interoperability.md

```diff
@@ -0,0 +1,277 @@
+
+OIDC also defines that the parameter acr_values is OPTIONAL and does not specify possible values. This leads to interoperability issues.
+
+This documents defines that Camara OpenId Providers MUST ignore the parameter acr_values. 
+
+This document defines that Camara Clients SHOULD not use the acr_values parameter. 
+
+> To foster interoperability a future version of this document might define values for the acr_values parameter acceptable in Camara.
+
+## Access Token Request
+
+This CAMARA document makes the scope parameter REQUIRED for the OAuth2 Client Credentials Grant.
+
+The client MUST authenticate with the authorization server as described in [Client Authentication Section](#client-authentication).
+
+## The Scope Parameter
+
+
+Scope values determine the specific CAMARA services being requested by the Service Provider, subject to the SP being registered to use those services. The scope values must be documented in the API OAS files by all Camara API subprojects. This document does not change OIDC definitions of scope values.
+
+
+---
+**NOTE**
+
+Scope values are an integral part of any OAuth2 and OIDC implementation. The RS enforces API access based on scope (if the Camara API subproject defines scopes).
+Therefore scopes should be available to API implementations.
+
+---
+
+
+## Missing "openid" scope
+
+[OIDC Core Authentication Request](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) states the following about the value of scope.
+> REQUIRED. OpenID Connect requests MUST contain the openid scope value. If the openid scope value is not present, the behavior is entirely unspecified. Other scope values MAY be present. ...
+
+This document defines that the authentication server SHOULD not return an id token if `openid` is missing in the scope parameter.
+
+This document defines the following error handling for a missing "openid" value in scope. 
+Please refer to [Authentication Error Response](https://openid.net/specs/openid-connect-core-1_0.html#AuthError).
+
+If "openid" is missing in the scope value but a claim that is [standardized in OIDC](https://openid.net/specs/openid-connect-core-1_0.html#Claims) is requested, then the Authorization Server returns an HTTP response code of 400 (Bad Request) and an error invalid_request.
+
+Clients SHOULD follow the OIDC standard and SHOULD include `openid` in the list of requested scopes.
+Without id token there is no `sub` field and the privacy features of OIDC are severely crippled. 
+Globally unique identifiers, like the MSISDN, should be avoided for privacy reasons.
+
+## Purpose
+
+A transaction specific request parameter purpose as specified in [openid-connect-4-identity-assurance-1_0-13](https://openid.net/specs/openid-connect-4-identity-assurance-1_0.html#name-transaction-specific-purpos) MUST be used to allow a SP to state the purpose for the transfer of End-User data it is asking for.
```

Comment:
I made a comment [here](https://github.com/camaraproject/IdentityAndConsentManagement/issues/140#issuecomment-2061425844) about this proposed text. My interpretation of the Identity Assurance specification is that the `purpose` request parameter is to be added as an extension of the [`claims`](https://openid.net/specs/openid-connect-core-1_0.html#Claims) request parameter (see [example](https://openid.net/specs/openid-connect-4-identity-assurance-1_0.html#section-6.1-11)).

But I don't think that is what is intended here. If it is what is intended, it needs to be far better explained.

Reply:
I like the simplicity of this "purpose" proposal. The reference to identity assurance is mostly something that says "others doing it the same way". 

I think the hardest part in standardization is finding a small meaningful step that people can agree on. It could all be standardized more fully with higher complexity, but then people never agree.
As TEF said, let's go with this for this release, and continue working on purpose the next release.

Reply:
@eric-murray it is important to note that the complete document introduces two different things which are both named `purpose`:

1. in chapter 11 (which is explicitly linked here in the text above) the transaction specific "**request parameter** `purpose` to allow an RP to state the purpose for the transfer of End-User data it is asking for"  
2. in chapter 6.1 an (optional) "additional **field** purpose to allow an RP to state the purpose for the transfer of a certain End-User Claim it is asking for" ... that is where you picked the examples from, but that is not meant here.

I suppose we are on the same stage with no 1., right? There is already the [request for examples](https://github.com/camaraproject/IdentityAndConsentManagement/issues/143) which would help here, but I understood that they will potentially go into another document.

Reply:
@hdamker 
> I suppose we are on the same stage with no 1., right?

So first, let me say that I do understand what is meant, I just think linking to the eKYC-IDA document introduces unnecessary ambiguity, and the possibility that someone who has not attended the ,meetings or read the comments would misinterpret what is meant.

> it is important to note that the complete eKYC-IDA document introduces two different things which are both named `purpose`:

I don't think the eKYC-IDA document explains this first use of `purpose` at all well. The term "request parameter" is doing a lot of heavy lifting there, which is not wise for a term that is not defined. Indeed, in [Section 6.1](https://openid.net/specs/openid-connect-4-identity-assurance-1_0.html#section-6.1-10), the specification refers to `purpose` as both a "field" and "parameter", so I'm not convinced that the authors are using separate definitions.

A reasonable interpretation of the term "request parameter" would be a "parameter included within the request", and not only "a query or form parameter". Or if I took my definition of "request parameter" from OAS, then that would exclude "parameters" in the request body. So I do not think we can rely on the use of the term "parameter" to ensure that __only__ the first meaning is assumed.

Linking to a document that defines two uses for a "variable" named `purpose` is surely a recipe for ambiguity. If all we are taking from the eKYC-IDA document is the name `purpose` and its format, then why not just state those explicitly in the OIDC profile and avoid this ambiguity?

Reply:
@AxelNennker 
> I like the simplicity of this "purpose" proposal. The reference to identity assurance is mostly something that says "others doing it the same way".

I couldn't find an existing implementation of OIDC that supports `purpose` as an additional query parameter for `/authorize` or additional form parameter for `/bc-authorize`, but there are some that support `purpose` as an additional field of `claims`.

Do you have an example of an authorisation server which has implemented the identity assurance specification in the way intended here?

Reply:
We agree on that Camara need some way to specify "purpose", right?
At the same time we/I want to keep vendor lock-in to a minimum.
Both approaches, purpose-parameter and purpose-coded-in-scope, require some special handling at the AZ.
I very much like to keep to standards, so at first I supported the purpose-coded-in-scope way but that approach also needs some special handling of the scopes at the AZ. I think, that purpose-coded-in-scope works in simple cases for of-the-shelf-OIDC-AZ but not for Camara/GSMA use cases. 

Now my understanding of the matter is that some custom code is needed at the AZ anyway and I think the purpose-parameter approach is simple, clear and moves us in the right direction. And also alread specified in eKYC-IDA.


> This specification introduces the request parameter purpose to allow an RP to state the purpose for the transfer of End-User data it is asking for.

> purpose: OPTIONAL. String describing the purpose for obtaining certain End-User data from the OP. The purpose MUST NOT be shorter than 3 characters and MUST NOT be longer than 300 characters. If these rules are violated, the authentication request MUST fail and the OP returns an error invalid_request to the RP.

By defining that the dpv-values must be used we are even clearer than eKYC-IDA.

Reply:
@AxelNennker
> We agree on that Camara need some way to specify "purpose", right?

Yes, and this is being discussed in #140. My understanding is that the current text for this PR is attempting to describe Option 1 of that issue. I'm happy with Option 1; my concern is that the current text proposal with the link to the eKYC-IDA document does not describe that option very well, primarily because it is not very well described in the eKYC-IDA document itself.

This is particularly true for CIBA, which does not have "request parameters", and can have either signed or unsigned flavours with different body formats. Whilst I can surmise what is required simply by replacing the current `scope` parameter in my requests with a `scope` and `purpose` pair of parameters at the same "level", that might not be obvious to any developers coming to this "fresh".

Some examples would help, of course, but a specification should not have to rely on examples for disambiguation.

Reply:
Please explain, what is not "well described". We all agree that Camara needs something-something-purpose. The initial proposal to encode purpose into scopes had limitations and also there is a disconnect between API subgroups who define "technical scopes" and purpose-encoded-scopes. So, we tried to replace the 0.1 version with the simple purpose parameter and dpv-values. I do not see, what is missing in the description. I think specifying a new parameter purpose and its dpv-values is so simple that is even does not require examples.
We are going to add examples later https://github.com/camaraproject/IdentityAndConsentManagement/issues/143. Maybe we add an implementers-guidelines document...

I see no way to misread the purpose proposal.

Here is the OIDC example from https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest + purpose:

> The following is a non-normative example HTTP 302 redirect response by the Client, which triggers the User Agent to make an Authentication Request to the Authorization Endpoint (with line wraps within values for display purposes only):

```javascript
HTTP/1.1 302 Found
  Location: https://server.example.com/authorize?
    response_type=code
    &scope=openid%20profile%20email
    &client_id=s6BhdRkqt3
    &state=af0ifjsldkj
    &redirect_uri=https%3A%2F%2Fclient.example.org%2Fcb
    &purpose=dpv:dpvValue
```

 If the AZ supports request objects then the field `purpose` is added to the request object. Similar CIBA.
```
{
   ...
   "purpose" : "dpv:dpvValue"
   ...
}

```

Is there a way to interpret the `purpose`request parameter differently? I don't see it.